src/main/java/mate.academy/lib/Injector.java

```diff
@@ -1,13 +1,93 @@
 package mate.academy.lib;
 
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+import java.util.HashMap;
+import java.util.Map;
+import mate.academy.service.FileReaderService;
+import mate.academy.service.ProductParser;
+import mate.academy.service.ProductService;
+import mate.academy.service.impl.FileReaderServiceImpl;
+import mate.academy.service.impl.ProductParserImpl;
+import mate.academy.service.impl.ProductServiceImpl;
+
 public class Injector {
     private static final Injector injector = new Injector();
+    private final Map<Class<?>, Object> instances = new HashMap<>();
 
     public static Injector getInjector() {
         return injector;
     }
 
-    public Object getInstance(Class<?> interfaceClazz) {
-        return null;
+    public Object getInstance(Class<?> clazz) {
+        if (instances.containsKey(clazz)) {
+            return instances.get(clazz);
+        }
+
+        try {
+            // Check if the class is an interface
+            if (clazz.isInterface()) {
+                Class<?> implementationClass = findImplementationClass(clazz);
+                Object instance = createInstance(implementationClass);
+                injectFields(instance);
+                instances.put(clazz, instance);
+                return instance;
+            } else {
+                Object instance = createInstance(clazz);
+                injectFields(instance);
+                instances.put(clazz, instance);
+                return instance;
+            }
+        } catch (Exception e) {
+            throw new RuntimeException("Failed to create an instance of class: "
+                    + clazz.getName(), e);
+        }
+    }
+
+    private Class<?> findImplementationClass(Class<?> interfaceClass) {
+        // Implement your logic to find the implementation class for the interface
+        // This is just a placeholder; you should customize it based on your application's
+        if (interfaceClass.equals(FileReaderService.class)) {
+            return FileReaderServiceImpl.class;
+        } else if (interfaceClass.equals(ProductParser.class)) {
+            return ProductParserImpl.class;
+        } else if (interfaceClass.equals(ProductService.class)) {
+            return ProductServiceImpl.class;
+        }
+        throw new RuntimeException("No implementation class found for: "
+                + interfaceClass.getName());
```

Comment:
what if we'd have 100500 intefaces? consider creating Map of interface to implementation key pair
```suggestion
        if (interfaceClazz.isInterface()) {
            return interfaceImplementations.get(interfaceClazz);
        }
        return interfaceClazz;
```